src/compiler/checker.ts

```diff
@@ -10403,6 +10433,7 @@ namespace ts {
                 case SyntaxKind.TemplateExpression:
                     return checkTemplateExpression(<TemplateExpression>node);
                 case SyntaxKind.StringLiteral:
+                    return checkStringLiteralExpression(<LiteralExpression>node);
                 case SyntaxKind.NoSubstitutionTemplateLiteral:

```

Comment:
Here's the most recent model:
```
    const valUndefined: BitSet = [2, BitValues.Undefined];
    const valNull: BitSet = [2, BitValues.Null];
    const valTrue: BitSet = [2, BitValues.True];
    const valFalse: BitSet = [2, BitValues.False];
    const valNaN: BitSet = [2, BitValues.NotNumber];

    const setStart = 2;

    let currentBitSet = setStart;
    let currentBitValue = 32;

    // Mapping of string values to bit sets
    // "a" -> [2, 32]
    // "b" -> [2, 64]
    // "c" -> [2, 128]
    // "d" -> [2, 256]
    // ...
    // "gg" -> [3, 1] // Set increments when reached 2^32 bits (32 values per set)
    let stringToBitSet: Map<BitSet> = {
        "null": valNull,
        "undefined": valUndefined
    };

    export function mapStringsToBitSets(values: string[], bitMap: BitSets = {}): BitSet[] {
        let bitSet: BitSet;
        for (let val of values) {
            if (stringToBitSet[val] === undefined) {
                if (currentBitValue === int32.highestBit) {
                    stringToBitSet[val] = [currentBitSet, currentBitValue];
                    currentBitValue = 1;
                    currentBitSet++;
                } else {
                    stringToBitSet[val] = [currentBitSet, currentBitValue];
                    currentBitValue = currentBitValue << 1;
                }
            }

            bitSet = stringToBitSet[val];
            if (bitMap[bitSet[0]] === undefined) {
                bitMap[bitSet[0]] = [bitSet[0], bitSet[1]];
            } else {
                bitMap[bitSet[0]][1] |= bitSet[1];
            }
        }
        let bitSets: BitSet[] = [];
        for (let x in bitMap) {
            bitSets.push(bitMap[x]);
        }
        return bitSets;
    }
```
So strings are never actually compared, only the ```set of bits```, most of the time there's only ```1 bitset``` (so a single integer check).